### PR TITLE
Introduce ZeROOffloadSubscriber for ORTModule

### PR DESCRIPTION
**Pull request by pengwa**
### Introduce ZeROOffloadSubscriber for ORTModule

As part of the work: integrate ORTModule with DeepSpeed stage3, this PR mainly focus on moving original PyTorch-based (leveraging hooks) param partition/offload implementation to ORTModule compatible implementation. 

Changes include:
1. Refactor `SubscriberBase`/`SubcriberManager` to support pre-forward/post_forward hooks. 
2. Implement new `ZeROOffloadSubscriber` by re-using DeepSpeed hook function as much as possible. Since all hook functions are defined in `DeepSpeedZeRoOffload._register_hooks_recursively` and `DeepSpeedZeRoOffload.setup_zero_stage3_hooks`, and the good thing is, the closure is not complex, all hooks are referencing the owning `DeepSpeedZeRoOffload` instance, so we can create new hook function with `FunctionType` by binding the owning `DeepSpeedZeRoOffload` instance, then call the new created function in subscriber's `pre_forward_module_apply_impl` and `post_forward_module_apply_impl` interfaces. 
3. Monkey patch `DeepSpeedZeRoOffload.setup_zero_stage3_hooks` to register the `ZeROOffloadSubscriber` for the model, then we don't need change any code on the DeepSpeed repo (at least so far). 
4. Fix the ATen embedding custom symbolic exporter function by tolerating weights size be (0) (changed by DeepSpeed zero stage 3). 

UT will be added once stage3 is fully supported.

### CODE REVIEW

File: orttraining/orttraining/python/training/utils/torch_type_map.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    "Undefined": [torch.onnx.TensorProtoDataType.UNDEFINED, None],
+}
+
+
+_DTYPE_TO_ONNX = {torch_dtype: onnx_dtype for k, (onnx_dtype, torch_dtype) in _CAST_PYTORCH_TO_ONNX.items()}
+
+
+def pytorch_dtype_to_onnx(dtype: torch.dtype) -> torch.onnx.TensorProtoDataType:
+    """Converts a pytorch dtype to an onnx dtype."""
+    return _DTYPE_TO_ONNX[dtype]
```
Comment (ajindal1) [inline orttraining/orttraining/python/training/utils/torch_type_map.py:37]: Should we add an edge case when the dtype is not present in the _DTYPE_TP_ONNX dict?

**Review by ajindal1 — APPROVED** [review]
LGTM

File: orttraining/orttraining/python/training/utils/torch_type_map.py (hunk 1 above)
Comment (pengwa) [reply orttraining/orttraining/python/training/utils/torch_type_map.py:39]: added check. thanks!

File: orttraining/orttraining/python/training/utils/torch_type_map.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+
+def pytorch_dtype_to_onnx(dtype: torch.dtype) -> torch.onnx.TensorProtoDataType:
+    """Converts a pytorch dtype to an onnx dtype."""
+    if dtype not in _DTYPE_TO_ONNX:
+        raise RuntimeError(f"Unsupported dtype {dtype}")
+    return _DTYPE_TO_ONNX[dtype]
+
+
+def pytorch_scalar_type_str_to_onnx(scalar_type: str) -> torch.onnx.TensorProtoDataType:
```
Comment (askhade) [inline orttraining/orttraining/python/training/utils/torch_type_map.py:42]: can this be covered as a case in "pytorch_dtype_to_onnx"? This way callers can just call "pytorch_dtype_to_onnx".

File: orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py [hunk 3]
```diff
@@ -659,6 +669,14 @@ def _log_feature_stats(self):
                 )
             )
 
+        feature_map.append(
+            (
+                "ZeRO Stage3 Compat",
```
Comment (askhade) [inline orttraining/orttraining/python/training/ortmodule/_graph_execution_manager.py:674]:
```suggestion
                "ZeRO Stage3 Support",
```

**Review by askhade — DISMISSED** [review]
LGTM

File: orttraining/orttraining/python/training/utils/torch_type_map.py (hunk 2 above)
Comment (pengwa) [reply orttraining/orttraining/python/training/utils/torch_type_map.py]: Good idea. Let me merge them into one function.

**Review by ajindal1 — APPROVED** [review]

**Pull request merged**